### PR TITLE
Don't query high cardinality labels.

### LINKED ISSUE #884: Don't query very high cardinality labels
(Building on an idea from #209)

We have a common class of query on very-high cardinality metrics (~800k timeseries) which actually only need to read a small number (~500) of those timeseries.  These queries usually involve multiple (4) matchers, 1-2 of them match pretty much all timeseries for that metric (like the job matcher), and 1-2 end up filtering it down to the 500 series in the end.

With the previous chunk store, this would pretty much always OOM, as it ended up reading millions of chunk descriptors before intersecting them away.  With the series store (#875), these queries no longer OOM, they're just slow, measures in 10s of seconds.

I propose we keep a in-querier-process cache of “matcher effectiveness”, measured as the number of series matched by a matcher that making it into a final result.  The idea would be that this would allow us to detect and ignore matchers which match lots of series that don’t survive the intersection.

Or perhaps alternatively we could keep a cache of observed label cardinality and just not read labels beyond a certain size (ignoring those matchers).  Eventually if we are forced to ignore all the matchers, we could fail the whole query with a cardinality error.

TLDR: both these proposal involve ignoring matchers and potentially reading more chunks, but avoiding doing very high cardinality queries on the index, saving memory and hopefully decreasing latency.

WDYT?

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #884 

This change limits the cardinality for labels; if the limit is exceeded, queries can still work, but there must be another selector with lower cardinality.

Notably, after this change, queries on two high-cardinality labels that would have results in a small number of series will fail.

Builds on seriesStores, so includes #875.

### CODE REVIEW

**Comment by tomwilkie** [comment]
After some testing, this improved latency for those particular queries mentioned in #884 by >2x.

File: pkg/chunk/chunk_store_test.go [hunk 1]
```diff
@@ -22,23 +22,28 @@ import (
...
 )
 
+type schemaFactory func(cfg SchemaConfig) Schema
+type storeFactory func(StoreConfig, Schema, StorageClient) (Store, error)
+
 var schemas = []struct {
 	name              string
-	fn                func(cfg SchemaConfig) Schema
+	schemaFn          schemaFactory
+	storeFn           storeFactory
```
Comment (bboreham) [inline pkg/chunk/chunk_store_test.go]: I was expecting this member to come in as part of the refactor to multiple stores.

File: pkg/chunk/chunk_store_test.go [hunk 2]
```diff
@@ -14,22 +14,45 @@ import (
...
 	"golang.org/x/net/context"
 
 	"github.com/weaveworks/common/test"
 	"github.com/weaveworks/common/user"
 )
 
+var schemas = []struct {
+	name              string
+	fn                func(cfg SchemaConfig) Schema
+	requireMetricName bool
```
Comment (bboreham) [inline pkg/chunk/chunk_store_test.go]: I expected `storeFn` to arrive in this commit, even though they would all be the same.

File: pkg/chunk/heap_cache.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
+package chunk
+
+import (
+	"container/heap"
+	"sync"
+	"time"
+)
+
+// heapCache is a simple string -> interface{} cache which uses a heap to manage
```
Comment (bboreham) [inline pkg/chunk/heap_cache.go]: Do we really need to write another cache?

File: pkg/chunk/heap_cache.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+// evictions.  O(log N) inserts, O(n) get and update.
```
Comment (bboreham) [inline pkg/chunk/heap_cache.go]: It looks like it should be O(1) get - is that a typo or am I missing something?

**Comment by tomwilkie** [comment]
Going to roll this into #875 as I'm struggling to keep track of all my open PRs.

**Pull request closed without merging**